test/lint/lint-filenames-permissions.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+This checks that all files in the repository have correct filenames and permissions
+"""
+
+import os
+import re
+import sys
+from subprocess import check_output
+
+CMD_ALL_FILES = "git ls-files --full-name"
+CMD_SOURCE_FILES = 'git ls-files --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
+CMD_SHEBANG_FILES = """git grep --full-name --line-number -I '^#!'"""
+ALLOWED_FILENAME_REGEXP = "^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$"
+ALLOWED_SOURCE_FILENAME_REGEXP = "^[a-z0-9_./-]+$"
+ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP = (
+    "^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)"
+)
+ALLOWED_PERMISSION_NON_EXECUTABLES = 644
+ALLOWED_PERMISSION_EXECUTABLES = 755
+ALLOWED_EXECUTABLE_SHEBANG = {
+    "py": [b"#!/usr/bin/env python3"],
+    "sh": [b"#!/usr/bin/env bash", b"#!/bin/sh"],
+}
+
+
+def check_all_filenames():
+    """
+    Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase
+    alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.
+    """
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    filename_regex = re.compile(ALLOWED_FILENAME_REGEXP)
+    exit_code = 0
+    for filename in filenames:
+        if not filename_regex.match(filename):
+            print(
+                f"""File "{filename}" does not not match the allowed filename regexp ('{ALLOWED_FILENAME_REGEXP}')."""
+            )
+            exit_code = 1
+    return exit_code
+
+
+def check_source_filenames():
+    """
+    Checks only source files (*.cpp, *.h, *.py, *.sh) against a stricter allowed regexp to make sure only lowercase
+    alphanumerics (a-z0-9), underscores (_), hyphens (-) and dots (.) are used in source code filenames.
+
+    Additionally there is an exception regexp for directories or files which are excepted from matching this regexp.
+    """
+    filenames = check_output(CMD_SOURCE_FILES, shell=True).decode("utf8").strip().split("\n")
```

Comment:
Out of curiosity: Would it make sense to replace `strip()` with `rstrip()` (https://docs.python.org/3/library/stdtypes.html?highlight=rstrip#str.rstrip)?

Just for fun: There is still one very specific case where this does not work (I hope I'm not missing something): You can add a file like this `git add test/util/zzz.py  ` (note the trailing whitespace) and `strip()` will make sure that `test/util/zzz.py` will be processed and not `test/util/zzz.py  `. I think your code is fine but if it can be improved wrt robustness I'm all in (possibly split first and remove trailing empty entries?)

Reply:
Great catch!

I've updated the logic to not use a `strip()` at all, and instead split on the newlines and then remove the trailing empty element. I chose to do this using a list comprehension filtering for empty list elements, rather than a `.pop()` of the last element, as it seemed slightly more robust, but can change it to a `pop()` if I'm overthinking it.

I tested that this does now catch files with trailing whitespace:
```bash
File "test/util/rpcauth-test.py " does not not match the allowed filename regexp ('^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$').
```